### ISSUE TITLE
URI / Location values in JSON 'can not be determined with webdriver.io' ?

### ISSUE THREAD

**Issue by raumB** [webdriverio-community/wdio-cucumberjs-json-reporter#60]
**Describe your question with as much detail as possible**
Hello, being not a specialist in this subject I just wanted to understand the reason for https://github.com/webdriverio-community/wdio-cucumberjs-json-reporter/blob/v6/lib/reporter.js#L238 and https://github.com/webdriverio-community/wdio-cucumberjs-json-reporter/blob/v6/lib/reporter.js#L319. Is this due to some principle technical restriction and it never would be possible to have the URI / location parameters implemented? Thanks and kind regards.

**If it's about a specific piece of code, try and include some of it to support your question.**
```
            uri: 'Can not be determined',
```
(https://github.com/webdriverio-community/wdio-cucumberjs-json-reporter/blob/v6/lib/reporter.js#L238)
```
                location: 'can not be determined with webdriver.io'
```
(https://github.com/webdriverio-community/wdio-cucumberjs-json-reporter/blob/v6/lib/reporter.js#L319)

**Environment (please complete the following information):**
 - **Node.js version:** 10.16.3
 - **NPM version:** 
 - **webdriver.io version:** ~6.11.x
 - **@wdio/cucumber-framework version:** ~6.11.1
 - **wdio-cucumberjs-json-reporter version:** ~2.0.2
 
**Additional context**
Cucumber reporters for other technologies seem to to offer these parameters which could be used for forward navigation to the test case.

**Comment by wswebcreation** [comment]
Hi @raumB 

WebdriverIO behaves in a different way regarding Cucumber reports in comparison to other frameworks that use CucumberJS. WebdriverIO has its own way of reporting and this module is translating that way back into JSON.
During the time I created this module there was no way to determine that, maybe there is now. That's why I added a label for people to help with this.

**Comment by christian-bromann** [comment]
This data should be available now as we updated the payload information with the last Cucumber release. @raumB you want to give us a hand and help us fix this?

**Comment by raumB** [comment]
Thanks a lot for the updates. I understand that in principle an enhancement with regard to the mentioned properties would be possible. With which exact versions of WDIO and Cucumber such an enhancement would work together? I'm asking since we are still using WDIO v6 for some other reason. Maybe we would need to go to v7 and maybe also switch in parallel to higher versions of other components? Thanks and kind regards.

**Comment by christian-bromann** [comment]
@raumB I would recommend to switch to v7 anyway. If you aren't a TypeScript user this update should be fairly simple. See more on this in our [migration guide](https://webdriver.io/docs/v7-migration). Then we can fix this and release a new version of the reporter that you can then use in your project. Happy to help in this process.

**Comment by raumB** [comment]
We plan to evaluate v7 for our test setup.

**Comment by raumB** [comment]
@christian-bromann @wswebcreation 
In an evaluation project we have now managed to migrate from WDIO v6 to v7 and would be able to test on v7 level.
To our understanding the URI parameter should contain some link or path to the corresponding feature file and the location parameter should contain some indicator for the specific corresponding test (for example, a line number from the feature file).
I assume there are some standards around about syntax and format of the `cucumber.json` file and what I'm suggesting for the two parameters should be checked against their common use in that file.
I myself could not contribute when it comes down to development activities, but I could try to test things, of course.
Thanks and kind regards.

**Comment by wswebcreation** [comment]
I've checked the output but the WDIO reporter is not giving back this information for the reporter. Will keep this one open, but this might first need to be fixed in the Cucumber framework adapter

**Comment by raumB** [comment]
Does also the WDIO v7 reporter does not give back this information?

**Comment by wswebcreation** [comment]
Nope, this is not given back, but maybe something Cucumber is giving back, but not passing it to the Cucumber framework adapter. Need to dive into this as this might be a 3rd party dependency, but if someone wants to take a look at it then be my guest

**Comment by silke-vormbrock** [comment]
@wswebcreation
Sorry for not coming back to your comment for such a long time.
We finally got the confirmation from development that the uri and the line for the feature file level would be sufficient to ensure traceability. Your support in this topic would be very much appreciated!

**Comment by silke-vormbrock** [comment]
@wswebcreation
would it be ok to invite for a meeting to discuss possible next steps?

**Comment by wswebcreation** [comment]
Hi @silke-vormbrock 

Sorry for not getting back to this. I have it on my to-do list, which is pretty long ATM :(, but will notify you when this is released

**Comment by silke-vormbrock** [comment]
Hi @wswebcreation 
thank you so much for coming back to me. Do you have a rough idea when this release can be expected? 
This missing piece would allow us to visualize fully automated traceability. Thanks!

**Comment by wswebcreation** [comment]
Hi @silke-vormbrock 

I hope before the end of the month

**Comment by silke-vormbrock** [comment]
Hi @wswebcreation,
that would be brilliant! As Product Excellence Award Winners 2021 we are invited to the devX Quality Talk on Feb 8 and one of our topics is how we ensure traceability in every pipeline run. Thank you!

**Comment by wswebcreation** [comment]
@silke-vormbrock 

A new version has been published, see https://github.com/webdriverio-community/wdio-cucumberjs-json-reporter/releases/tag/v4.2.0

This will now add the `line` and `uri`, see the example below

```json
[
  {
    "keyword": "Feature",
    "type": "feature",
    "description": "",
    "line": 2,
    "name": "Sample",
    "uri": "/Users/Git/wdio-cucumber-v6-v7/v7/test/features/sample.feature",
    "tags": [
      {
        "location": {
          "line": 1,
          "column": 1
        },
        "name": "@sample",
        "id": "4"
      }
    ],
    "elements": [
      {
        "keyword": "Scenario",
        "type": "scenario",
        "description": "",
        "name": "cukes",
        "tags": [
          {
            "name": "@sample",
            "astNodeId": "4"
          }
        ],
        "id": "sample;cukes",
        "steps": [
          {
            "arguments": [],
            "keyword": "Before",
            "name": "",
            "result": {
              "status": "passed",
              "duration": 1000000
            },
            "line": null,
            "match": {
              "location": "can not be determined with webdriver.io"
            }
          },
          {
            "arguments": [],
            "keyword": "Before",
            "name": "",
            "result": {
              "status": "passed",
              "duration": 0
            },
            "line": null,
            "match": {
              "location": "can not be determined with webdriver.io"
            }
          },
          {
            "arguments": [],
            "keyword": "Given",
            "name": "I have a background",
            "result": {
              "status": "passed",
              "duration": 2000000
            },
            "line": null,
            "match": {
              "location": "can not be determined with webdriver.io"
            }
          },
          {
            "arguments": [],
            "keyword": "Given",
            "name": "I have 42 cukes in my belly",
            "result": {
              "status": "passed",
              "duration": 2905000000
            },
            "line": null,
            "match": {
              "location": "can not be determined with webdriver.io"
            }
          },
          {
            "arguments": [],
            "keyword": "After",
            "name": "",
            "result": {
              "status": "passed",
              "duration": 1000000
            },
            "line": null,
            "match": {
              "location": "can not be determined with webdriver.io"
            }
          }
        ]
      }
    ],
    "id": "sample",
    "metadata": {
      "browser": {
        "name": "chrome",
        "version": "97.0.4692.71"
      },
      "device": "Test Device",
      "platform": {
        "name": "ios",
        "version": "14"
      }
    }
  }
]

```

Can you please confirm and then close this issue if this fixes your needs

**Comment by silke-vormbrock** [comment]
Hi @wswebcreation,

thank you so much! I forwarded this information to development and hope to get feedback very quickly.

**Comment by silke-vormbrock** [comment]
Hi @wswebcreation,
our developer is checking the solution currently, we expect his results by the end of the week.

**Comment by wswebcreation** [comment]
Cool, thanks for the update

**Comment by silke-vormbrock** [comment]
Hi @wswebcreation,
unfortunately, we are not yet there where we want to be, but your update helped us to move into the right direction. I scheduled a meeting with our colleagues from Cumulus and will come up with a final update after this meeting. Thank you for your effort!

**Comment by silke-vormbrock** [comment]
Hi @wswebcreation,
after consuming your correction, this information is available in our Pipeline Dashboard in Sirius:
![image](https://user-images.githubusercontent.com/96289297/152323187-5b0f324b-0ce7-4c74-9eb5-10693b6217a4.png)
This link behind "can not be determined with webdriver.io" is taking the user to the correct repository but the written information is misleading. Can you support us to get a better text here?
Only with your correction we finally have the information in column "Source Path" what is required to find the correct test case. Our Cumulus colleagues are checking currently how to "crop" this information because only the part of the source path starting with feature/... is necessary.
Thank you for checking ;-)

**Comment by wswebcreation** [comment]
Hi @silke-vormbrock 


`can not be determined with webdriver.io` is being determined per step and not per test case/scenario. So I wonder how your reporting tool works. Can you explain that a bit more?

**Comment by silke-vormbrock** [comment]
Hi @wswebcreation,
this is the answer I got from the developer who supports us in this topic:
Hi Wim, you are correct at the moment each step information is taken to display the test case. This might change in the future to the scenario name. To drill down into the scenario it could be useful to link the individual steps to its implementation. Would this be possible via webdriver.io?

**Comment by silke-vormbrock** [comment]
Hi Wim,
thank you so much for your support. Our dev colleagues took over for the next steps. 
Please allow me to keep this GitHub issue open until we have the final solution in place.
Thanks! Best regards, Silke

**Comment by wswebcreation** [comment]
Sure, no problem

**Comment by christian-bromann** [comment]
Any updates @silke-vormbrock ?

**Comment by jgrammen-agilitypr** [comment]
any chance on reviving (now that we are on wdio v9) this issue and getting it over the finish line? it would be nice to have the per step statistics available.

**Comment by christian-bromann** [comment]
@jgrammen-agilitypr as the labels suggest we appreciate all types of contributions that help to resolve this issue.

**Comment by jgrammen-agilitypr** [comment]
in the `beforeScenario` hook, based on https://stackoverflow.com/questions/64627229/how-can-i-access-the-file-name-a-test-runs-in-from-inside-the-aftertest-hook-in, it is possible to get the line number and column number
```
        const feature = world.gherkinDocument.feature;
        const docChildren = feature?.children;

            // Find the scenario and step
            let scenario;
            let step;
            for (const docChild of docChildren) {
            let possibleScenario = docChild.scenario ? docChild.scenario : docChild.background;
            // Find the step in the scenario
            for (const possibleStep of possibleScenario.steps) {
                step = possibleStep;
                scenario = possibleScenario;
                break;
            }
            if (step) {
                break;
            }
            }
            if (!scenario || !step) {
            console.log("Error is coming from an unknown location in the gherkin feature file.");
            } else {
            console.log(`Error found in ${feature?.name}:`);
            console.log(`\t${scenario.keyword}: ${scenario.name} (${scenario.location.line}:${scenario.location.column})`);
            console.log(`\tStep: ${step.keyword} ${step.text} (${step.location.line}:${step.location.column})`);
            }
```
